### ISSUE TITLE
Compiling in Visual Studio 2015

### ISSUE THREAD

**Issue by talmo** [kaspermarstal/SimpleElastix#15]
Compiling under Windows 8.1 x64 with the current version of Visual Studio 2015 Community results in some issues having to do with the HDF5 library in VTK. I've reproduced this on two Win8.1 machines now.

This issue is present in both the current [SimpleElastix](https://github.com/kaspermarstal/SimpleElastix/tree/f19998a830f8dbe341391db8d154b3636b1e8b9f) and [development](https://github.com/kaspermarstal/SimpleElastix/tree/f082285a3e5706180920038c3d0bc3064e844298) branches.

First, here's the relevant snippet of the build log:
```
2>  -- Checking for appropriate format for 64 bit long:
2>  Width with l64 failed with result: 255
2>  Width with l failed with result: 255
2>  Width with L failed with result: 255
2>  Width with q failed with result: 255
2>  Width with I64 failed with result: 255
2>  -- Checking for apropriate format for 64 bit long: not found
2>  Width with ll failed with result: 255
```

Here are the popup dialog error boxes that VS2015 throws:
```
---------------------------
Microsoft Visual C++ Runtime Library
---------------------------
Debug Assertion Failed!

Program: ...F5\src\itkhdf5\CMake\CMakeFiles\CMakeTmp\Debug\cmTC_7951f.exe
File: minkernel\crts\ucrt\inc\corecrt_internal_stdio_output.h
Line: 2480

Expression: ("Invalid integer length modifier", 0)

For information on how your program can cause an assertion
failure, see the Visual C++ documentation on asserts.

(Press Retry to debug the application)

---------------------------
Abort   Retry   Ignore   
---------------------------




---------------------------
Microsoft Visual C++ Runtime Library
---------------------------
Debug Assertion Failed!

Program: ...F5\src\itkhdf5\CMake\CMakeFiles\CMakeTmp\Debug\cmTC_3ff09.exe
File: minkernel\crts\ucrt\inc\corecrt_internal_stdio_output.h
Line: 2480

Expression: ("Invalid integer length modifier", 0)

For information on how your program can cause an assertion
failure, see the Visual C++ documentation on asserts.

(Press Retry to debug the application)

---------------------------
Abort   Retry   Ignore   
---------------------------



---------------------------
Microsoft Visual C++ Runtime Library
---------------------------
Debug Assertion Failed!

Program: ...F5\src\itkhdf5\CMake\CMakeFiles\CMakeTmp\Debug\cmTC_bcca7.exe
File: minkernel\crts\ucrt\inc\corecrt_internal_stdio_output.h
Line: 2480

Expression: ("Invalid integer length modifier", 0)

For information on how your program can cause an assertion
failure, see the Visual C++ documentation on asserts.

(Press Retry to debug the application)

---------------------------
Abort   Retry   Ignore   
---------------------------



---------------------------
Microsoft Visual C++ Runtime Library
---------------------------
Debug Assertion Failed!

Program: ...F5\src\itkhdf5\CMake\CMakeFiles\CMakeTmp\Debug\cmTC_7a9dc.exe
File: minkernel\crts\ucrt\inc\corecrt_internal_stdio_output.h
Line: 2480

Expression: ("Invalid integer length modifier", 0)

For information on how your program can cause an assertion
failure, see the Visual C++ documentation on asserts.

(Press Retry to debug the application)

---------------------------
Abort   Retry   Ignore   
---------------------------



---------------------------
Microsoft Visual C++ Runtime Library
---------------------------
Debug Assertion Failed!

Program: ...F5\src\itkhdf5\CMake\CMakeFiles\CMakeTmp\Debug\cmTC_0a437.exe
File: minkernel\crts\ucrt\inc\corecrt_internal_stdio_output.h
Line: 2480

Expression: ("Invalid integer length modifier", 0)

For information on how your program can cause an assertion
failure, see the Visual C++ documentation on asserts.

(Press Retry to debug the application)

---------------------------
Abort   Retry   Ignore   
---------------------------



---------------------------
Microsoft Visual C++ Runtime Library
---------------------------
Debug Assertion Failed!

Program: ...F5\src\itkhdf5\CMake\CMakeFiles\CMakeTmp\Debug\cmTC_229e1.exe
File: minkernel\crts\ucrt\inc\corecrt_internal_stdio_output.h
Line: 2480

Expression: ("Invalid integer length modifier", 0)

For information on how your program can cause an assertion
failure, see the Visual C++ documentation on asserts.

(Press Retry to debug the application)

---------------------------
Abort   Retry   Ignore   
---------------------------
```

After tracking this issue down, it seems that [this was a VTK issue back from Aug 2015](http://www.vtk.org/pipermail/vtkusers/2015-August/091842.html) (see this [merge](https://gitlab.kitware.com/vtk/vtk/merge_requests/446)).

On the ITK side, this fix was incorporated in [ITK 4.8.2](https://blog.kitware.com/announcing-itk-4-8-2/) (see [this commit](http://itk.org/gitweb?p=ITK.git;a=commit;h=b7f44755e9de335bfeb43187d327c3536db207e4)).

The issue is with SimpleITK though. SimpleElastix currently uses SimpleITK 0.9.1 which was built on ITK 4.8.1 (see [release log](http://www.itk.org/Wiki/SimpleITK/ReleaseNotes#SimpleITK_-_Version_0.9.1_Release)).

The [master branch](https://github.com/SimpleITK/SimpleITK/tree/master) of SimpleITK actually already has support for ITK 4.9.0 which was [released on Feb 3, 2016](https://blog.kitware.com/kitware-releases-itk-4-9-to-further-rapid-evolution-in-medical-imaging/) (see the [changes in master since the SimpleITK 0.9.1 release](https://github.com/SimpleITK/SimpleITK/compare/v0.9.1...master)).

I'm not sure how to change the CMake configuration to use the newest version of the master branch of SimpleITK instead of the v0.9.1 label (or if that's even desirable) but a quick fix to enable error-free compiling in Visual Studio 2015 involves just changing the ITK version from 4.8.1 to 4.8.2, which I think is less likely to break things in SimpleElastix than 4.9.0.

This involves just changing [SuperBuild/External_ITK.cmake:L38](https://github.com/kaspermarstal/SimpleElastix/blob/develop/SuperBuild/External_ITK.cmake#L38) from:
`set(ITK_TAG_COMMAND GIT_TAG v4.8.1)`
To:
`set(ITK_TAG_COMMAND GIT_TAG v4.8.2)`

I chose not to just fix and submit a pull request for this since I wasn't sure about the whole SimpleITK release label business. Also my build is still compiling so I can't confirm that this fix alone is enough to resolve the issue and doesn't break things.

Let me know if there's a better fix!

**Comment by kaspermarstal** [comment]
Hi talmo, thank you for the extensive bug report AND possible solution! Let me know if your problem is solved with ITK v4.8.2 and the SimpleElastix branch will be updated accordingly. You are also welcome to submit a PR if you prefer. Don't worry about the versioning.

**Comment by talmo** [comment]
Okay so, the issues above don't seem to crop up anymore, but now there appears to be a new problem.

After building from ALL_BUILD, most things get compiled until we get to SimpleITK, at which point Visual Studio throws these errors:
```
Severity	Code	Description	Project	File	Line	Suppression State
Error	C2061	syntax error: identifier 'MovingMaskType' [C:\se\SimpleITK-build\Code\Elastix\src\SimpleElastix.vcxproj]	SimpleITK	c:\toolbox\simpleelastix\code\elastix\src\sitkSimpleElastix.hxx	37	
Error	MSB6006	"cmd.exe" exited with code 1.	SimpleITKExamples	C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\V140\Microsoft.CppCommon.targets	171	
Error	C2039	'FixedMaskType': is not a member of 'elastix::ElastixFilter<TFixedImage,TMovingImage>' [C:\se\SimpleITK-build\Code\Elastix\src\SimpleElastix.vcxproj]	SimpleITK	c:\toolbox\simpleelastix\code\elastix\src\sitkSimpleElastix.hxx	32	
Error	C2061	syntax error: identifier 'FixedMaskType' [C:\se\SimpleITK-build\Code\Elastix\src\SimpleElastix.vcxproj]	SimpleITK	c:\toolbox\simpleelastix\code\elastix\src\sitkSimpleElastix.hxx	32	
Error	C2039	'MovingMaskType': is not a member of 'elastix::ElastixFilter<TFixedImage,TMovingImage>' [C:\se\SimpleITK-build\Code\Elastix\src\SimpleElastix.vcxproj]	SimpleITK	c:\toolbox\simpleelastix\code\elastix\src\sitkSimpleElastix.hxx	37	
Error	C2061	syntax error: identifier 'MovingMaskType' [C:\se\SimpleITK-build\Code\Elastix\src\SimpleElastix.vcxproj]	SimpleITK	c:\toolbox\simpleelastix\code\elastix\src\sitkSimpleElastix.hxx	37	
Error	C2039	'FixedMaskType': is not a member of 'elastix::ElastixFilter<TFixedImage,TMovingImage>' [C:\se\SimpleITK-build\Code\Elastix\src\SimpleElastix.vcxproj]	SimpleITK	c:\toolbox\simpleelastix\code\elastix\src\sitkSimpleElastix.hxx	32	
Error	C2061	syntax error: identifier 'FixedMaskType' [C:\se\SimpleITK-build\Code\Elastix\src\SimpleElastix.vcxproj]	SimpleITK	c:\toolbox\simpleelastix\code\elastix\src\sitkSimpleElastix.hxx	32	
Error	C2039	'MovingMaskType': is not a member of 'elastix::ElastixFilter<TFixedImage,TMovingImage>' [C:\se\SimpleITK-build\Code\Elastix\src\SimpleElastix.vcxproj]	SimpleITK	c:\toolbox\simpleelastix\code\elastix\src\sitkSimpleElastix.hxx	37	
```
Re: paths on this setup -- the source repo is in `C:\Toolbox\SimpleElastix` and the build directory is in `C:\se` since I was having issues with long paths earlier.

Here's a [gist with the full error log](https://gist.github.com/talmo/7c388cf2f59f23bb8334).

So here's some initial digging: this is on f082285 which is where `Code/Elastix/src/sitkSimpleElastix.hxx` got changed to use: `typename ElastixFilterType::FixedMaskType*` as the fixed mask type.

Visual Studio seems to be picking this up as there are a few errors referencing this file/line, e.g.:
```
4>c:\toolbox\simpleelastix\code\elastix\src\sitkSimpleElastix.hxx(32): error C2039: 'FixedMaskType': is not a member of 'elastix::ElastixFilter<TFixedImage,TMovingImage>' [C:\se\SimpleITK-build\Code\Elastix\src\SimpleElastix.vcxproj]
4>            with
4>            [
4>                TFixedImage=ImageType1,
4>                TMovingImage=ImageType2
4>            ]
```

Could this be a bug or am I doing something wrong?

**Comment by kaspermarstal** [comment]
Are you on the SimpleElastix or develop branch?

**Comment by talmo** [comment]
I'm on develop -- specifically commit f082285, which is 3 commits behind the current HEAD but it doesn't seem like any of them (3d7755f, 1dd325a or a4db5bc) contain any changes pertinent to this issue.

**Comment by kaspermarstal** [comment]
You are seeing this error because you are living on the bleeding edge :grin: the elastix trunkpublic is out of sync with develop. You can try to go back one commit to 4412307 (but I cannot guarantee it will work) or use the SimpleElastix branch.

Develop contains a completely rewritten elastix core and must be considered extremely unstable. You are welcome to continue using this branch of course, but FYI there are still issues related to the use of masks and 4D groupwise registration is not supported yet.

**Comment by kaspermarstal** [comment]
This is actually the same problem you are experiencing in #16. There are changes to elastix trunkpublic on my machine that have not been published yet.

**Comment by talmo** [comment]
Yup, I just noticed. Mk, I'll close both of these until there's an update on elastix trunkpublic. I'm currently on the latest (r5219) for reference.

I did get the Windows version compiling (and testing the Ubuntu one now) after rolling back that commit above AND fixing a couple of errors in `sitkSimpleElastix.hxx` having to do with not returning `*this` in some of the parameter functions.

I'll keep an eye out for the mask fixes :)
